pkg/gktest/result.go

```diff
@@ -1,14 +1,62 @@
 package gktest
 
+import (
+	"errors"
+)
+
 // Result is the structured form of the results of running a test.
-type Result struct{}
+type Result interface {
+	// String formats the Result for printing.
+	String() string
+	// IsFailure returns true if either a test failed or there was a problem
+	// executing tests.
+	IsFailure() bool
+
+	Equal(Result) bool
+}
+
+type Success struct{}
+
+var _ Result = &Success{}
+
+func (s *Success) String() string {
+	panic("implement me")
+}
+
+func (s *Success) IsFailure() bool {
+	panic("implement me")
+}
+
+func (s *Success) Equal(right Result) bool {
+	success, isSuccess := right.(*Success)
+	if !isSuccess {
+		return false
+	}
+	return *s == *success
+}
+
+func errorResult(err error) *ErrorResult {
+	return &ErrorResult{err}
+}
+
+type ErrorResult struct {
+	error
+}
+
+func (r *ErrorResult) Equal(right Result) bool {
+	err, isErr := right.(*ErrorResult)
+	if !isErr {
+		return false
+	}
+
+	// errors.Is is not commutative, but Equal should be.
+	return errors.Is(r.error, err.error) || errors.Is(err.error, r.error)
```

Comment:
Equal() is required to make unit tests via cmp.Diff sane. If Equal make it not commutative, then cmp.Diff will no longer be commutative and it will be confusing to future test authors unless they always know to follow the correct cmp.Diff(want, got) order. If we are comfortable with assuming authors will always know to put want before got for cmp.Diff, then it is fine.

Choosing not to define Equal() is even worse, as then it'd mean every test involving Result types would need to define its own logic for this. Or we'd need test fixtures just for comparing these, which would do effectively the same thing as cmp.Diff but be annoying to maintain and not provide as pretty output.

Reply:
This has been resolved by using a new builtin cmpopts that is compatible with go 1.13 errors.